index/scorch/snapshot_rollback_test.go

```diff
@@ -88,6 +108,21 @@ func TestIndexRollback(t *testing.T) {
 		t.Fatal(err)
 	}
 
+	rollbackPointsB, err := sh.RollbackPoints()
+	if err != nil || len(rollbackPointsB) != 3 {
```

Comment:
how does len(rollbackPoints) go from 1 to 3 when we only executed one more batch?

Reply:
> how does len(rollbackPoints) go from 1 to 3 when we only executed one more batch?

That's a great question... a head scratcher that needed a bunch of debugging printf's.

The reason why seems to be because of that issue where the persister loop seems runs twice for each introduced segment, as the persister loop is tracking a lastPersistedEpoch that's not quite right.  That is, that "lastPersistedEpoch" might be better renamed or thought of as "the epoch of the snapshot that we last examined for persistence", which is not the same as the epoch # that persistSnapshot() ends up using.

That double-looping should be fixed, imho, but I'd aim or recommend instead that that should be a separate PR.

Reply:
(Also, another weirdness or mystery is I'm wondering why travis-CI didn't catch any (?) rollback test failures before.)

Reply:
I think it's time I finally understood this "double looping" thing you keep complaining about.

I'm now starting to think that it's actually coming from the fact that the persister now introduces it's own new epoch, which it necessarily sees again later (but it should largely do nothing, so it was deemed harmless).  Is this in fact the double looping you're talking about? Because if so, it's nothing something we can easily make go away without rearchitecting the whole thing.

Reply:
Well no one ever filed a bug with the error, it's just speculative fixes that the test "must be racey".  What we do know is that they were quite stable from the last time Abhinav adjusted them.  Like they ran a lot the whole month of January without issue.  So, I suspect something we did exposed a problem.

Reply:
> Is this in fact the double looping you're talking about? Because if so, it's nothing something we can easily make go away without rearchitecting the whole thing

Yep, that's it!  And, I also sense it's not necessarily a big rearchitecting... perhaps instead along the lines of persistSegment() needs to return the epoch number that it just assigned and persisted.  And, then the persisterLoop can check whether that matches where it was last was plus 1.  If it's not last+1, then that means the introducer concurrently snuck some segments into the mix in the meanwhile (so next we should jump back up to the top of the persisterLoop to start persisting those).

> Well no one ever filed a bug with the error, it's just speculative fixes that the test "must be racey". What we do know is that they were quite stable from the last time Abhinav adjusted them. Like they ran a lot the whole month of January without issue. So, I suspect something we did exposed a problem.

Yep -- I'm thinking all the optimizations that came in recently must have changed timings somehow... but, along the way, I was thinking "oh, it's just weird me", and "it's some test raciness that affects me somehow" -- but I should've filed a bug.  Then, @sreekanth-cb also mentioned this weekend that he was also seeing a problem with rollback test failure, so the rabbit ears perked up more.

Reply:
yep. I started seeing this issue in my local and I was expecting the CI to be failed for my open PRs.. But for some reason it didn't..